### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by ancr7**

### CODE REVIEW

File: .idea/.gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
```
Comment (hardikbohra) [inline .idea/.gitignore:1]: add below folders in the same file to ignore:
.gradle/ 
.idea/
build/

File: src/main/java/Assignment1.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
+import java.util.Scanner;
+
+class ItemList {
```
Comment (hardikbohra) [inline src/main/java/Assignment1.java]: create Model class for ItemList (to follow MVC)

File: src/main/java/Assignment1.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    String _name, _type;
+    double _price;
+    int _quantity;
+
+    // method to extract word from a sentence from pos till end or '-' appears
+    String extractWord(String input_string, int pos) {
```
Comment (hardikbohra) [inline src/main/java/Assignment1.java]: Move all extraction and calculation logic to service class. (to follow MVC)

Comment (hardikbohra) [inline src/main/java/Assignment1.java]: follow variable naming convention (Ex : input_string should be written as inputString)

File: src/main/java/Assignment1.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+            System.out.println("Invalid quantity");
+            return false;
+        }
+        // price should only have numbers and can have decimal (.) once.
+        if (!price.matches("[0-9.]*") || (price.length() - price.replace(".", "").length() > 1)) {
+            System.out.println("Invalid price");
+            return false;
+        }
+        // type check.
+        if (!type.equals("raw") && !type.equals("manufactured") && !type.equals("imported")) {
```
Comment (hardikbohra) [inline src/main/java/Assignment1.java]: use enum instead for type

File: src/main/java/Assignment1.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            System.out.println("Invalid type");
+            return false;
+        }
+        return true;
+    }
+
+    // method to extract data from input to member variables
+    void extractData(String input_string) {
+        String name = extractWord(input_string, input_string.indexOf("-name") + 5).trim();
```
Comment (hardikbohra) [inline src/main/java/Assignment1.java]: create constants for -name, -quantity, -price, -type as these strings are used multiple times in the business logic as well.

File: src/main/java/Assignment1.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        String type = extractWord(input_string, input_string.indexOf("-type") + 5).trim();
+        _name = name;
+        _quantity = Integer.parseInt(quantity);
+        _price = Double.parseDouble(price);
+        _type = type;
+    }
+
+    // method to calculate raw tax.
+    double rawTax(double total) {
+        if (total < 0) throw new RuntimeException("Total cant be -ve");
```
Comment (hardikbohra) [inline src/main/java/Assignment1.java]: what if total is equal to 0?

File: src/main/java/Assignment1.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+        else if (total <= 200)
+            tax += 10;
+        else
+            tax += (total + tax) * 0.05;
+        return tax;
+    }
+
+    // method to calculate total cost.
+    double calculateCost() {
+        double totalCost = _price * _quantity;
```
Comment (hardikbohra) [inline src/main/java/Assignment1.java]: no need to start with _ for private member variables.

File: src/test/java/ItemListTest.java [hunk 8]
```diff
@@ -0,0 +1,222 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class ItemListTest {
+
+    @Test
```
Comment (hardikbohra) [inline src/test/java/ItemListTest.java]: create @Before annotation to do the setup before executing any testcase.
You can move ItemList initialization in the setup method.
follow : 
https://junit.org/junit4/javadoc/4.12/org/junit/Before.html

File: src/test/java/ItemListTest.java [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+    @DisplayName("When input is Empty")
+    void checkIfInputEmpty() {
+        ItemList item = new ItemList();
+        assertEquals(false,item.isInputValid(""));
```
Comment (hardikbohra) [inline src/test/java/ItemListTest.java]: with assertion, you can pass message also for user readability.
follow : 
http://junit.sourceforge.net/javadoc/org/junit/Assert.html#assertArrayEquals(java.lang.String,%20byte[],%20byte[])

File: src/test/java/ItemListTest.java [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+        });
+
+        String expectedMessage = "Invalid input";
+        String actualMessage = exception.getMessage();
+
+        assertTrue(actualMessage.contains(expectedMessage));
+    }
+
+    @Test
+    void rawTaxSampleTest1() {
```
Comment (hardikbohra) [inline src/test/java/ItemListTest.java]: don't use Test1, Test2 naming convention.

**Review by hardikbohra — CHANGES_REQUESTED** [review]

File: .idea/.gitignore (hunk 1 above)
Comment (ancr7) [reply .idea/.gitignore:1]: Done.

File: src/main/java/Assignment1.java (hunk 3 above)
Comment (ancr7) [reply src/main/java/Assignment1.java]: done.

File: src/main/java/Assignment1.java (hunk 6 above)
Comment (ancr7) [reply src/main/java/Assignment1.java]: I think total can be 0 if price is zero.

File: src/main/java/Assignment1.java (hunk 7 above)
Comment (ancr7) [reply src/main/java/Assignment1.java]: done.

File: src/test/java/ItemListTest.java (hunk 10 above)
Comment (ancr7) [reply src/test/java/ItemListTest.java]: done.

File: src/main/java/Assignment1.java (hunk 4 above)
Comment (ancr7) [reply src/main/java/Assignment1.java]: done.

File: src/main/java/Assignment1.java (hunk 2 above)
Comment (ancr7) [reply src/main/java/Assignment1.java]: done.

File: src/main/java/Assignment1.java (hunk 3 above)
Comment (ancr7) [reply src/main/java/Assignment1.java]: done.

**Pull request merged**